webpack/components/Table/TableHooks.js

```diff
@@ -41,14 +42,143 @@ class ReactConnectedSet extends Set {
       this.delete(id);
     }
   }
+
+  addAll(ids) {
+    ids.forEach(id => super.add(id));
+    this.forceRender();
+  }
 }
 
-const useSet = (initialArry) => {
+export const useSet = (initialArry) => {
   const [, setToggle] = useState(0);
   // needed because mutating a Ref won't cause React to rerender
   const forceRender = () => setToggle(prev => prev + 1);
   const set = useRef(new ReactConnectedSet(initialArry, forceRender));
   return set.current;
 };
 
-export default useSet;
+export const useSelectionSet = (results, metadata, initialArry = [], idColumn = 'id') => {
+  const selectionSet = useSet(initialArry);
+  const ids = results?.map(result => result[idColumn]) ?? [];
+  const areAllRowsOnPageSelected = () => Number(ids?.length) > 0 &&
+                                         ids.every(result => selectionSet.has(result));
+
+  const selectPage = () => selectionSet.addAll(ids);
+  const selectNone = () => selectionSet.clear();
+  const onRowSelect = (isSelected, id) => {
+    if (isSelected) {
+      selectionSet.add(id);
+    } else {
+      selectionSet.delete(id);
+    }
+  };
+
+  const selectedCount = selectionSet.size;
+
+  const isSelected = id => selectionSet.has(id);
+
+  return {
+    onRowSelect,
+    selectedCount,
+    areAllRowsOnPageSelected,
+    selectPage,
+    selectNone,
+    isSelected,
+    selectionSet,
+  };
+};
+
+export const useBulkSelect = (results, metadata, initialArry = [], idColumn = 'id') => {
+  const { selectionSet: inclusionSet, ...selectOptions } =
+                useSelectionSet(results, metadata, initialArry, idColumn);
+  const exclusionSet = useSet([]);
+  const [searchQuery, updateSearchQuery] = useState('');
+  const [isSelectAllChecked, setSelectAll] = useState(false);
+  const selectedCount = isSelectAllChecked ?
+    Number(metadata.subtotal) - exclusionSet.size : selectOptions.selectedCount;
+
+  const areAllRowsOnPageSelected = () => isSelectAllChecked ||
+                                         selectOptions.areAllRowsOnPageSelected();
+
+  const isSelected = (id) => {
+    if (isSelectAllChecked) {
+      return !exclusionSet.has(id);
+    }
+    return inclusionSet.has(id);
+  };
+
+  const selectPage = () => {
+    setSelectAll(false);
```

Comment:
What if the total number of items happens to be the number of items on the page?

Reply:
If you chose `selectPage` you clearly revoking the idea of `selectAll` and instead doing `selectAll on this page`
setSelectAll is purely for marking the `selectAll mode` since I need that to keep track of the `exclusionSet`.

`selectPage` is still using only the `inclusionSet`

Reply:
Gotcha.  Yeah, per my previous comment I think it's a good idea to rename that `selectAllMode` then.